### PR TITLE
Expand [status - crowd controlled] trigger

### PR DESCRIPTION
**Pull request by asaka-wa**
# Description

The current CC trigger is entirely binary. People often want to know if they're controlled by a specific CC type for which they have a counter. 

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings

### CODE REVIEW

**Comment by Nightwarden24** [comment]
- Could you add C_LossOfControl.GetActiveLossOfControlDataCount and C_LossOfControl.GetActiveLossOfControlData to  to [.luacheckrc](https://github.com/WeakAuras/WeakAuras2/blob/master/.luacheckrc#L721) file?

- I guess there are no these functions in Classic.
   Yeah, they have different names
   
   
   I can't say anything about structure of return value

- List of loss_of_control_Types is not complete. What about at least these?

   + BANISH (Cyclone - Druid; Imprison if use Detainment PvP-talant - Demon hunter)
   + DISORIENT (Dragon's Breath - Mage; Scatter Shot - Hunter; Blind - Rogue)
   + INCAPACITATE (Repentance - Paladin; Paralysis - Monk?; Gouge - Rouge; Incapacitating Roar - Druid)
   + POLYMORPH (Polymorph - Mage; Hex - Shaman)
   + SLEEP (Song of Chi-Ji - Monk; Hibernate - Druid)

   Does the function return data for them? Or am I completely wrong about return type for control abilities? And existence of types needs to be checked in Classic  ¯\\\_(ツ)_/¯
   
   P.S. Not sure about these types
   + SAP (Sap - Rouge [SAP or INCAPACITATE?]; Imprison without PvP-talent - Demon Hunter [SAP or INCAPACITATE?])
   + SNARE (Slow - Mage; Thunder Clap - Warrior)
   + HORROR (Mortal Coil - Warlock; Vampiric Touch - Priest {If it's dispelled, the dispeller flees in Horror for 3 sec})

**Comment by mrbuds** [comment]
>     * I guess there are no these functions in Classic.

This trigger is disabled on classic https://github.com/WeakAuras/WeakAuras2/blob/master/WeakAuras/Prototypes.lua#L7851

**Comment by Nightwarden24** [comment]
> > ```
> > * I guess there are no these functions in Classic.
> > ```
> 
> This trigger is disabled on classic https://github.com/WeakAuras/WeakAuras2/blob/master/WeakAuras/Prototypes.lua#L7851

I apologize for that

**Comment by asaka-wa** [comment]
There's no API or table of types on Blizzard's side so I could certainly be missing some types. I took the list I'm using from an addon. I know that Hex, for example, is actually a "PACIFYSILENCE" effect with the displayText "Polymorphed" so just because some types you might intuitively think must exist aren't listed that doesn't mean they're missing.

**Comment by Nightwarden24** [comment]
Ah, I thought you checked types manually (at least the most typical abilities or "ambiguous" cases) :disappointed:

I didn't rely on intuition or display text. I was looked at mechanic of abilities on wow.tools and wowhead.com and thought it had something to do with locType :flushed:

**Comment by asaka-wa** [comment]
My guess is that it's like the debuffTypes on auras, and is more about what spells can remove the effects than it is about what the effects themselves are.

**Comment by mrbuds** [comment]
I found this with a google search that lead me to <https://www.curseforge.com/wow/addons/d4hp/issues/5>
```
AllowedLOCTypes = <table> {
1 = "STUN_MECHANIC"
2 = "STUN"
3 = "PACIFYSILENCE"
4 = "SILENCE"
5 = "FEAR"
6 = "CHARM"
7 = "PACIFY"
8 = "CONFUSE"
9 = "POSSESS"
10 = "SCHOOL_INTERRUPT"
11 = "ROOT"
12 = "FEAR_MECHANIC"
}
```

**Comment by asaka-wa** [comment]
Yeah, I handle all of those.

**Comment by mrbuds** [comment]
I see, mhm i would have preferred re-affect the value of locType, and make it's condition of type "select" instead of "string"

**Comment by asaka-wa** [comment]
To be fair, the controlType value is using the "select" conditionType so locType can probably just have its conditionType removed. 
One would be the version as they select it, while the other would include those combined ones like "PACIFYSILENCE" but I don't think that's really going to be relevant.

File: WeakAuras/Types.lua
Comment (InfusOnWoW) [inline WeakAuras/Types.lua:1202]: lowercase "T"

File: WeakAuras/Prototypes.lua
Comment (InfusOnWoW) [inline WeakAuras/Prototypes.lua:6684]: Just "Name"

File: WeakAuras/Prototypes.lua
Comment (InfusOnWoW) [inline WeakAuras/Prototypes.lua:6693]: Just "Spell Name"

File: WeakAuras/Prototypes.lua
Comment (InfusOnWoW) [inline WeakAuras/Prototypes.lua:6702]: Just "Type"

File: WeakAuras/Prototypes.lua
Comment (InfusOnWoW) [inline WeakAuras/Prototypes.lua:6711]: Just "Spell Id"

File: WeakAuras/Prototypes.lua
Comment (InfusOnWoW) [inline WeakAuras/Prototypes.lua:6716]: operator_types == "only_equal"

File: WeakAuras/Prototypes.lua
Comment (InfusOnWoW) [inline WeakAuras/Prototypes.lua:6631]: Did you check what the lookoutSchool is if a spell with multiple spell schools is interrupted?

E.g. frostfire, chaos bolt or whatever else tehre might be?

File: WeakAuras/Prototypes.lua
Comment (InfusOnWoW) [inline WeakAuras/Prototypes.lua:6730]: The init here should be unnecessary.

File: WeakAuras/Prototypes.lua
Comment (InfusOnWoW) [inline WeakAuras/Prototypes.lua:6737]: The init here should be unnecessary.

File: WeakAuras/Prototypes.lua
Comment (InfusOnWoW) [inline WeakAuras/Prototypes.lua:6743]: The init here should be unnecessary.

File: WeakAuras/Prototypes.lua
Comment (InfusOnWoW) [inline WeakAuras/Prototypes.lua:6757]: You could write that as "inverse ~= show", which is a bit less readable. Though it's fine in this case.

File: WeakAuras/Prototypes.lua
Comment (InfusOnWoW) [inline WeakAuras/Prototypes.lua:6722]: I think this should not be available as a  condition check.

File: WeakAuras/Prototypes.lua
Comment (InfusOnWoW) [inline WeakAuras/Prototypes.lua:6705]: You don't have to use init here.

File: WeakAuras/Prototypes.lua
Comment (InfusOnWoW) [inline WeakAuras/Prototypes.lua:6696]: You don't have to init here.

File: WeakAuras/Prototypes.lua
Comment (InfusOnWoW) [inline WeakAuras/Prototypes.lua:6672]: you set conditionType twice

**Comment by nullKomplex** [comment]
What is the difference between interrupted school and interrupt school?

File: WeakAuras/Prototypes.lua
Comment (asaka-wa) [reply WeakAuras/Prototypes.lua:6705]: How does the value get stored without it?

File: WeakAuras/Prototypes.lua
Comment (asaka-wa) [reply WeakAuras/Prototypes.lua:6631]: Hmm, yeah I'll see if I can test this later.

Comment (Nightwarden24) [reply WeakAuras/Prototypes.lua:6631]: I checked: for example, lockoutSchool is [124](https://wow.gamepedia.com/COMBAT_LOG_EVENT#Spell_School) when interrupting [Chaos Bolt](https://www.wowhead.com/spell=116858)

Comment (asaka-wa) [reply WeakAuras/Prototypes.lua:6631]: Thanks for checking. 
That would mean that they would be locked out of all [Fire, Nature, Frost, Shadow, Arcane], right?
So if I aim to provide a dropdown for the base school lockouts I need to associate the higher level school values with each of the base types. Or, actually looks like I can go via a bitmask.

File: WeakAuras/Prototypes.lua
Comment (InfusOnWoW) [reply WeakAuras/Prototypes.lua:6705]: It should work with just store = true if you have a local variable of that name.

Comment (asaka-wa) [reply WeakAuras/Prototypes.lua:6705]: Ho! nice.

File: WeakAuras/Prototypes.lua
Comment (Nightwarden24) [reply WeakAuras/Prototypes.lua:6631]: Another example: lockoutSchool is [72](https://wow.gamepedia.com/COMBAT_LOG_EVENT#Spell_School) when interrupting [Full Moon](https://www.wowhead.com/spell=274283)
Yes, all my abilities of nature and arcane school are locked out =(

Comment (asaka-wa) [reply WeakAuras/Prototypes.lua:6631]: Yup `bit.band(data.lockoutSchool, interruptSchool)` should handle it. I'll test tonight to make sure.

**Comment by asaka-wa** [comment]
Okay, changes made 👍

**Comment by Nightwarden24** [comment]
I'm totally against your list of control types. So I've checked class abilities that affect control, but have not considered spells from other sources. Result in file.

Interesting:

- Slow, knockback, taunt don't count as loss of control (captain obvious) and, accordingly, are not returned by the function.
- Events fired when using some abilities have more than one locType. Classic example: priest's [Mind Bomb](https://www.wowhead.com/spell=205369) (FEAR_MECHANIC and ROOT). I don't know whether you consider and handle it.
- displayText is just value from LOSS_OF_CONTROL_DISPLAY_X variable (Global string), not a localized string returned by client or server.

I got the following list of locTypes:

+ CHARM
+ CONFUSE
+ DISARM
+ FEAR
+ FEAR_MECHANIC
+ NONE - I don't know what to do with that. Ignore?
+ PACIFY
+ PACIFYSILENCE
+ POSSESS
+ ROOT
+ SCHOOL_INTERRUPT
+ SILENCE
+ STUN
+ STUN_MECHANIC

Types shouldn't be combined or omitted. They are too different by its nature. Don't take their names literally. I'll try to explain how I see logic of their assignment

* CHARM - ? I don't know
* CONFUSE - You are incapacitated and walking in a circle (i.e. there are small movements). Any damage will instantly interrupt the effect.
* DISARM - You can't use weapon or shield, but can still attack with hands
* FEAR - It's same as FEAR_MECHANIC, but lasts significantly less time (2-3 seconds)
* FEAR_MECHANIC - You are incapacitated and fleeing in an uncontrolled direction. Damage may cancel the effect.
* PACIFY - You can't attack, but can cast spell and move?
* PACIFYSILENCE - You can't attack or cast any spell, but can still move
* POSSESS - You can't control your actions. Someone else does it.
* ROOT -  You can't move, but can attack or cast spell
* SCHOOL_INTERRUPT - You can't cast only spell of specific school
* SILENCE - You can't cast any spell, but can move and attack
* STUN - You are incapacitated and don't move. Damage will break the effect or you are invulnerable 
* STUN_MECHANIC - Any stun. You are incapacitated except using special spell (example [Enraged Regeneration](https://www.wowhead.com/spell=184364)). Damage won't cancel effect.

Сheck out the file and see what abilities belong to each category. Perhaps, you will also notice the difference between them.

Your loss_of_control_types list will generate such questions:

- I want to track situation when I'm stunned and should use Dispersion to reduce incoming damage. But icon appers when Cyclone or Paralysis is applied.
- More exotic. Why does aura trigger on [Hex](https://www.wowhead.com/spell=51514)? I can't deal physical damage (i.e. attack) when I'm silenced? Opposite situation. I want to track situation when I can't attack and I'm silenced (there is bunch of other spells with this type of control), but aura is active when on [Garrote](https://www.wowhead.com/spell=703).

I suggest the following option:

+ CHARM - no idea, Charm?
+ CONFUSE = Confuse or Disorientation (closer to spell description for me)
+ DISARM = Disarm
+ FEAR = Horror
+ FEAR_MECHANIC = Fear
+ PACIFY = Pacify
+ PACIFYSILENCE - don't know
+ POSSESS = Possess
+ ROOT = Root
+ SCHOOL_INTERRUPT = Interrupt
+ SILENCE = Silence
+ STUN = Incapacitation
+ STUN_MECHANIC = Stun

P.S. Multiple select in cases of control types and spell schools would be really convenient.

File: [Loss of control.xlsx](https://github.com/WeakAuras/WeakAuras2/files/5660637/Loss.of.control.xlsx)

**Comment by asaka-wa** [comment]
Well, FEAR_MECHANIC is new to me, I'll incorporate that. I would've handled it in the same was as STUN, rolling both together but I don't really care about that, if anyone thought that they were important to be separate then, sure, whatever. 

PACIFYSILENCE though, sure could be handled separately – my code only gets easier with this change, but it feels like if people are under a PACIFY effect they want to know about PACIFYSILENCE too. 

All these names are too abstract for users to just *get*, as you've showed above it isn't clear even to us. WA sometimes just throws API names at people and lets them sink or swim, but that only works in oft-used parts of the API where collectively the user community interacts with it enough to just learn the way to refer to things. In this case I think that it would just mystify people and they'd just go ahead and pick "STUN", for example, and hope it does what they need. 

So, I dunno. I could accept your advice and simplify the trigger. I'm just not sure that best serves the users.

**Comment by Nightwarden24** [comment]
To pick stun, user has to select STUN_MECHANIC. STUN isn't a stun. Sorry, I don't know how else to explain this to you.

Give users all control types. Don't make decisions for them.

By the way, you still haven't added "new" functions to [.luacheckrc](https://github.com/WeakAuras/WeakAuras2/blob/master/.luacheckrc#L721) file.

**Comment by asaka-wa** [comment]
Sure, thanks for your thoughts. I'll bear them in mind.

**Comment by InfusOnWoW** [comment]
@Nightwarden24 are you happy with asakawa's latest changes?

File: WeakAuras/Prototypes.lua
Comment (InfusOnWoW) [inline WeakAuras/Prototypes.lua:6653]: Change to "Specific Type".

**Comment by Nightwarden24** [comment]
Good. Filling the list with types exactly as they are is not a bad idea. But I'm afraid that some description will still need to be added later.

One thing before the merge that you forgot about again: luacheck.

**Comment by asaka-wa** [comment]
Unless something official can be found then I see nothing to be gained by adding some ad hoc guesswork. 

I'm not forgetting, I have no intention of doing it. It's not a system I know and I'm sure the addon's maintainers can adjust that if need be.

**Review by mrbuds — APPROVED** [review]

**Pull request merged**
